activity/journal.go

```diff
@@ -0,0 +1,99 @@
+package activity
+
+import (
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// Journal is used to describe a log of activity. That activity is associated with a user-generated
+// event by the ID, and can include a description and start time to provide further context when
+// listed in activity history.
+type Journal struct {
+	ID, Description string
+	Start           time.Time
+}
+
+var mux = sync.Mutex{}
+
+func initHistory() ([]Journal, map[string]int) {
+	return make([]Journal, 0), make(map[string]int)
+}
+
+// history: a list of commands that have been invoked
+// storedJournals: lookup to identify whether a journal ID has already been recorded in history.
+var history, storedJournals = initHistory()
+
+// addToHistory appends the command description to history if it hasn't been registered before.
+func (j Journal) addToHistory() {
+	// Return if already added to history.
+	if _, ok := storedJournals[j.ID]; ok {
+		return
+	}
+
+	// Not yet set. Lock to make sure we only register the command once.
+	mux.Lock()
+	defer mux.Unlock()
+
+	// Check again in case something else beat us to it.
+	if _, ok := storedJournals[j.ID]; ok {
+		return
+	}
+
+	// Register the command.
+	storedJournals[j.ID] = len(history)
+	history = append(history, j)
+}
+
+// Record writes a new entry to the journal. It creates a new file for the journal if needed, then
+// appends the message to that journal. Journals are stored in the user's cache directory under
+// `wash/activity/ID.log`.
+func (j Journal) Record(msg string, a ...interface{}) {
```

Comment:
Does this need to be exported?

Reply:
No, but I don't see a reason not to. It seems like a valid part of normal operations for a Journal, and if for some reason you have direct access to it you might as well use it.

It might need some syncing on the printf operation though, I'm not sure that's thread-safe.